url46a88cc932_toby1984_tinyc_tgz-pJ8-de_codesourcery_dcpu16_compiler_CompilerJ8/src/de/codesourcery/dcpu16/parser/Lexer.java

```diff
@@ -0,0 +1,433 @@
+package de.codesourcery.dcpu16.parser;
+
+import java.io.EOFException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Stack;
+
+import de.codesourcery.dcpu16.ast.NumberLiteralNode;
+import de.codesourcery.dcpu16.compiler.Identifier;
+import de.codesourcery.dcpu16.compiler.Operator;
+import de.codesourcery.dcpu16.exceptions.ParseException;
+import de.codesourcery.dcpu16.util.IInputStreamProvider;
+
+public class Lexer
+{
+    private final IInputStreamProvider input;
+    private Scanner scanner;
+    
+    private final List<Token> tokens = new ArrayList<>();
+    private boolean atEOF = false;
+    
+    private boolean ignoreWhitespace = true;
+    
+    // the token returned by the last read() call , 
+    // required to be able to reset the scanner
+    // when switching from "ignore whitespace" -> "do not ignore whitespace"
+    private Token previousToken = null;
+    
+    // transient
+    private final StringBuilder buffer = new StringBuilder();
+    private final Stack<State> stack = new Stack<>();
+    
+    protected final class State 
+    {
+        private final List<Token> tokens = new ArrayList<>(Lexer.this.tokens);
+        private final boolean atEOF = Lexer.this.atEOF;
+        private final int currentOffset = scanner.getOffset();
+        private final boolean ignoreWhitespace = Lexer.this.ignoreWhitespace;
+        private final Token previousToken = Lexer.this.previousToken;
+        
+        public void apply() throws IOException 
+        {
+            Lexer.this.tokens.clear();
+            Lexer.this.tokens.addAll(tokens);
+            Lexer.this.atEOF = atEOF;
+            Lexer.this.ignoreWhitespace = ignoreWhitespace;
+            Lexer.this.previousToken = previousToken;
+            
+            final InputStream newIn = createInputStream( currentOffset );
+            Lexer.this.scanner = new Scanner( newIn, currentOffset);
+        }
+    }
+    
+    private InputStream createInputStream(int offset) throws EOFException, IOException {
+        return advanceInputStream( input.createInputStream() , offset );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code might contain a resource leak. A resource leak can slow down or crash your system.

**Fix**
Consider closing the resource returned by the *createInputStream* method in a try-finally block. Currently, there are execution paths that do not contain closure statements, for example, when an if condition at line 66 evaluates to true.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).